asyncpg/protocol/codecs/base.pyx

```diff
@@ -190,16 +193,29 @@ cdef class Codec:
             FastReadBuffer elem_buf = FastReadBuffer.new()
 
         elem_count = <ssize_t><uint32_t>hton.unpack_int32(buf.read(4))
+        if elem_count != len(self.element_type_oids):
+            raise OutdatedTypeCacheError(
+                'unexpected number of attributes of composite type {!r}.{!r}: '
```

Comment:
You're using `{!r}` for `self.schema` and `self.name`.  Wouldn't it render them in quotes, like this "... of composite type 'abc'.'def': ..."?

Reply:
Right. It is intended to avoid ambiguity when either `schema` or `name` has dots. It is similar to PG's `format('%I.%I', schema, name)` but I don't mind quotes type (single or double) and their presence in some cases.

It is just convenient format for parsing.

Reply:
Please don't rely on parsing when you can avoid it.  Store the data as attributes on the exception.  Postgres' native exceptions have the `schema` and `data_type_name` attributes for similar purposes.

Reply:
Here's a list of attribute names that asyncpg errors can have.  I'd try to use only them.

https://github.com/MagicStack/asyncpg/blob/59e2878bef959cfab35ebd19c211477c1daf1101/asyncpg/exceptions/_base.py#L26-L45

Reply:
Great! I almost forgot it.
I hope the `OutdatedTypeCacheError` should neither be inherited from the `PostgresMessage` class nor have the `PostgresMessageMeta` as a metaclass.

In the other case we would face troubles of filling the `severity` and `sqlstate` fields which are required.

Reply:
It should not inherit from `PostgresMessage`, since it does not originate from the server.  But it still makes sense to keep it close in shape to server errors.  With respect to the attributes, let's stick to just `schema` and `data_type_name` (that's what you were intending to parse out of the message anyway).

Reply:
@elprans you mean `schema_name`?

Reply:
Yes, `schema_name`.

Reply:
@elprans, thank you! I just wanted be sure I understand @1st1's hint correctly.
I'm sorry, I can push the discussed improvements/fixes only at the weekend.